Add CodeMirror support to Rules tab with syntax highlighting, line number, search and replace...

gunir opened the pull request:
### What does this PR do?

CodeMirror is a versatile, modern text editor component built specifically for the web, designed to handle complex editing tasks that go far beyond the capabilities of a standard HTML <textarea>. While a plain textarea is part of the basic browser toolkit, CodeMirror 6 (the version we implemented) is a complete rewrite focused on modularity, accessibility, and high performance.

Here is a comparison between CodeMirror and the previous plain textarea used in your Vite/Wails project:

Performance and Rendering

>    Virtualization: The standard Vite textarea renders the entire text buffer at once. If you have 10,000 lines of rules, the browser must manage all 10,000 lines in memory and layout. CodeMirror uses virtualization, meaning it only creates DOM elements for the few dozen lines currently visible on your screen, making scrolling instant regardless of file size.

>    Update Cycle: Your previous Vite setup used a "controlled component" pattern where every keystroke forced React to re-process the entire string, causing the lag you observed. CodeMirror manages its own internal state and only notifies React when changes are finished, bypassing the expensive React render loop.

Editing Features

>    Syntax Highlighting: A plain textarea only supports a single color for all text. CodeMirror allows for sophisticated parsing, which we used to highlight your uBlockOrigin comments in gray and syntax symbols in distinct colors.

>    Developer Tools: CodeMirror includes built-in support for professional features that a textarea lacks, such as native line numbers, specialized search-and-replace panels, and bracket matching.

>    Code-Aware Logic: Unlike the plain textarea, CodeMirror understands the document structure. This allowed us to implement the "Format Rules" feature to programmatically sort and deduplicate your rules while maintaining the integrity of your comments.

Layout and Feel

>    Scroll Behavior: The previous textarea was subject to browser-level "smooth scrolling" and momentum interpolation, which contributed to a sluggish feel in the Wails webview. CodeMirror provides direct control over the scroll container, allowing us to force an "instant" response to mouse and keyboard input.

>    Isolation: CodeMirror uses CSS containment to ensure that internal changes to the text do not trigger expensive layout recalculations for the rest of your Wails application UI.

### How did you verify your code works?

Check this image:

<img width="1498" height="1139" alt="image" src="https://github.com/user-attachments/assets/d501197d-ca65-45a4-aca8-15e69dfa0bcd" />


### What are the relevant issues?

No, just improvement

gunir committed:
Update index.tsx - Add CoreMirror support and search and replace

This PR add CodeMirror support to improve performance, improve rule editing experience

Signed-off-by: Gunir <134402102+gunir@users.noreply.github.com>

gunir committed:
Update index.css - Add CodeMirror

Update index.css - Add CodeMirror

Signed-off-by: Gunir <134402102+gunir@users.noreply.github.com>

anfragment reviewed (changes requested):
Please see the failing builds and the [frontend linter's output](https://github.com/ZenPrivacy/zen-desktop/actions/runs/20390790234/job/58620933456?pr=559) (automatically fixable with `npm run lint -- --fix`).

gunir reviewed (commented):
I see, I'll look into it later!

gunir committed:
Update index.css - Add rules loading

.rules--loading {
  display: flex;
  align-items: center;
  justify-content: center;
  height: 100%;
  width: 100%;
}

Signed-off-by: Gunir <134402102+gunir@users.noreply.github.com>

gunir committed:
Update index.tsx

Signed-off-by: Gunir <134402102+gunir@users.noreply.github.com>

gunir committed:
Update index.tsx - linted

Signed-off-by: Gunir <134402102+gunir@users.noreply.github.com>

gunir committed:
Update index.tsx

Signed-off-by: Gunir <134402102+gunir@users.noreply.github.com>

gunir committed:
Update index.tsx

Signed-off-by: Gunir <134402102+gunir@users.noreply.github.com>

gunir committed:
Update index.tsx

Signed-off-by: Gunir <134402102+gunir@users.noreply.github.com>

anfragment reviewed (changes requested):
@gunir Please take a look at the remaining linting and build errors.

The PR should also include the updated `package.json` and `package-lock.json`.
